File: vimo_python/vimo/communication.py
import sys
import requests
import pickle
import os
import pandas as pd
import platform
import re
from io import StringIO

def upload(model, model_name, model_desc, tags, train_dataset, train_dataset_name, dataset_desc, requirements_file, user_name, password):
	"""Function uploads scikit-learn model, the training set and all needed metadata to the vimo base.

	Parameters
	----------
	model : scikit-learn model or string
		model object or path to the model pickle
	model_name : string
		name of the model that will be visible in the vimo
	model_desc : string
		description of the model
	tags : list
		list of tags
	train_dataset : array-like or string
		array-like or path to csv file (must contain '/') or hash of already uploaded data, structure X|Y is required
	train_datset_name : string
		name of the dataset that will be visible in the vimo
	dataset_desc : string
		description of the dataset
	requirements_file : string
		path to python style requirements file, can be easily obtained by running: "pip freeze > requirements.txt" at your command line
	user_name : string
		your user name
	password : string 
		your password

	Returns
	-------
	string
		Returns an information if uploading the model was successful.
	"""

	# url to post
	url = 'http://192.168.137.64/models/post'

	# collecting system info
	if platform.system() == 'Linux':
		info = {'system': platform.system(),
			'system_release': platform.release(),
			'distribution': platform.linux_distribution()[0],
			'distribution_version': platform.linux_distribution()[1],
			'language': 'python',
			'language_version': platform.python_version(),
			'architecture': platform.architecture()[0],
			'processor': platform.machine()}

	info['model_name'] = model_name
	info['train_data_name'] = train_dataset_name


	# init of flag if train_dataset is a hash
	info['train_dataset_hash'] = 0

	# init of flags what temporary file should be removed at the end of function
	del_model = False
	del_train_data = False

	# uploading model
	if type(model) == str:
		# case when model is a path
		files = {'model': open(model, 'rb')}
	else:
		# case when model is an object

		# creating temporary file
		with open('.tmp_model.pkl', 'wb') as fd:
			pickle.dump(model, fd)

		# uploading model
		files = {'model': open('.tmp_model.pkl', 'rb')}

		# setting flag
		del_model = True

	# creating regexp to findout if the train_dataset is a path or id
	reg = re.compile("/")

	# uploading train dataset
	if type(train_dataset) == str and reg.search(train_dataset) is None:
		# case when train_dataset is a hash of already uploaded dataset

		info['train_dataset_hash'] = train_dataset

	elif type(train_dataset) == str:
		# case when train_dataset is a path to dataset

		files['train_dataset'] = open(train_dataset, 'rb')

	else:
		# case when train_dataset is a matrix

		# conversion to pandas data frame
		train_dataset = pd.DataFrame(train_dataset)

		# creating temporary file
		train_dataset.to_csv('./tmp_train_data_csv', index = False)

		# uploading dataset
		files['train_dataset'] = open('./tmp_train_data_csv', 'rb')

		# setting flag
		del_train_data = True

	if type(model_desc) == str and reg.search(model_desc) is not None:
		info['model_desc'] = open(model, 'rb').read()
	elif type(model_desc) == str:
		info['model_desc'] = model_desc

	if type(dataset_desc) == str and reg.search(dataset_desc) is not None:
		info['dataset_desc'] = open(model, 'rb').read()
	elif type(dataset_desc) == str:
		info['dataset_desc'] = dataset_desc

	# uploading requirements file
	files['requirements'] = open(requirements_file, 'rb')

	# setting session info flag
	info['is_sessionInfo'] = 0

	# user_name
	info['user_name'] = user_name

	# password
	info['password'] = password

	# tags
	info['tags'] = tags

	# creating request
	r = requests.post(url, files = files, data = info)

	# removing temporary files
	if del_model:
		os.remove('.tmp_model.pkl')
	if del_train_data:
		os.remove('./tmp_train_data_csv')

	return r.text

def predict(model_name, X, pred_type = 'exact', prepare_columns = True):
	"""
	Function uses model in the database to make a prediction on X.

	Parameters
	----------
	model_name : string
		name of the model in the base that you want to use
	X : array-like/string
		array-like or path to csv file (must containt '/') or hash of already uploaded dataset,
		must have column names if prepare_columns is set to False
	pred_type : string
		type of the prediction: exact/prob
	prepare_columns : boolean
		if true and if X is an object then take column names from model in the database

	Returns
	-------
	array-like
		Returns a pandas data frame with made predictions.
	"""

	# url
	url = 'http://192.168.137.64/models/' + model_name + '/predict/' + pred_type

	# regexp to find out if X is a path
	reg = re.compile("/")

	# uploading data
	if type(X) == str and reg.search(X) is None:
		# case when X is a hash
		info = {'is_hash': 1}
		info['hash'] = X

		# request
		r = requests.get(url, data = info)
	elif type(X) == str:
		# case when X is a path

		info = {'is_hash': 0}
		files = {'data': open(X, 'rb')}

		# request
		r = requests.get(url, files=files, data=info)
	else:
		# case when X is an object

		# conversion to pandas data frame
		X = pd.DataFrame(X)

		if prepare_columns:
			columns = model_info(model_name)['data_info']['columns']
			c = []
			for col in columns:
				c.append(col[1])
			c = c[0:-1]
			X.columns = c

		# creating temporary file
		X.to_csv('./tmp_data_csv-' + model_name, index = False)

		info = {'is_hash': 0}
		files = {'data': open('./tmp_data_csv-' + model_name, 'rb')}

		# request
		r = requests.get(url, files=files, data = info)

		# removing temporary file
		os.remove('./tmp_data_csv-' + model_name)

	return pd.read_csv(StringIO(r.text), header=None)

def model_info(model_name):
	"""
	Function model_ino

	...::: DESCRIPTION :::...
	Take all metadata about the model from vimo. It will return a dict.

	...::: PARAMS :::...
	model_name - name of the model in the vimo base

	...::: RETURN :::...
	Dict with fields:
		model_info
		data_info
	"""
	r = requests.get('http://192.168.137.64/models/' + model_name + '/info')
	return r.json()

def create_user(user_name, password, mail):
	"""
	Function create_userFunction creates new user in the vimo base.

	Parameters
	----------
	user_name : string
		your user name, has to be unique, if such user already exists you will get such information in response
	password : string
		your password, passwords are hashed
	mail : string
		your mail

	Returns
	-------
	string
		Information if creating account was successful.
	"""
	r = requests.post('http://192.168.137.64/users/create_user', data = {'user_name': user_name, 'password': password, 'mail': mail})
	return r.text

def search_model(row=None, column=None, missing=None, classes=None, owner=None, tags=None):
	"""
	Search vimo base for models with specific tags. If all parameters are set to None, then returns all models' name in vimo.

	Params
	------
	row : string
		string describing how many rows should have training dataset
		examples:
			row = '<101;'; row = '>120;'; row = '>100;<200;'
	column : string
		string describing how many columns should have training dataset
		examples:
			column = '<101;'; column = '>120;'; column = '>100;<200;'
	missing : string
		string descibing how many missing values should have training dataset
		examples:
			missing = '=0;'; missing = '>120;'; missing = '<10001;'; missing = '>100;<200;'
	classes : string
		string descibing how many classes should have training dataset
		examples:
			classes = '=2;'; classes = '<3;'; classes = '>2;'; classes = '>2;<11;'
	owner : string
		owner's user name
	tags : list
		list of tags, all should be strings

	Returns
	-------
	list
		Returns a list of models' names that have at least one common tag with that provided in parameter 'tags'
	"""
	data = {'row': row, 'column': column, 'missing': missing, 'classes': classes, 'owner': owner, 'tags': tags}
	r = requests.get('http://192.168.137.64/models/search', data=data)
	return r.json()['models']

File: vimo_r/vimo/R/communication.r
library(httr)

predict <- function(model_name, X, pred_type = 'exact') {
	# url
	url = paste0('http://192.168.137.64/models/', model_name, '/predict/', pred_type)

	body = list()

	del = FALSE

	# uploading data
	if(class(X) == "character" && !grepl("/", X)) {
		# case when X is a hash
		body[['is_hash']] =  1
		body[['hash']] = X

		# request
		POST(url = url, body = body)
	} else if(class(X) == "character") {
		# case when X is a path
		body[['is_hash']] =  0
		body[['data']] = upload_file(X)
	} else {
		# case when X is an object
		# creating temporary file
		write.table(X, paste0('./tmp_data_csv-', model_name), row.names=F, col.names=T, sep=',')

		body[['is_hash']] =  0
		body[['data']] = upload_file(paste0('./tmp_data_csv-', model_name))

		# removing temporary file
		del = TRUE
	}

	r = content(POST(url = url, body = body), as='text')

	if(del) {
		file.remove(paste0('./tmp_data_csv-', model_name))
	}
	read.csv(text = r, header = F)[,1]
}

upload <- function(model, model_name, train_dataset, train_dataset_name, test_dataset = NA, test_dataset_name = NA) {
	ses = sessionInfo()
	pkg = c(ses$otherPkgs, ses$loadedOnly)

	pkg_names = rep('a', length(pkg))
	pkg_versions = rep('a', length(pkg))

	for(i in 1:(length(pkg))) {
		pkg_names[i] = names(pkg[i])
		pkg_versions[i] = pkg[[i]]$Version
	}

	requirements = data.frame(pkg_names, pkg_versions)
	write.table(requirements, '.tmp-requirements.txt', row.names=F, col.names=F, sep=',')

	system = Sys.info()[1]
	system_release = Sys.info()[2]
	distribution = strsplit(ses[[4]], split = ' ')[[1]][1]
	distribution_version = strsplit(ses[[4]], split = ' ')[[1]][2]
	language = 'r'
	language_version = paste0(sessionInfo()[[1]]$major, '.', sessionInfo()[[1]]$minor)
	architecture = strsplit(ses[[1]]$system, ', ')[[1]][1]
	processor = Sys.info()[5]

	if(is.na(test_dataset)) {
		is_test_dataset = 0
	} else {
		is_test_dataset = 1
	}

	is_sessionInfo = 1

	del_model = F
	del_train_data = F
	del_test_data = F

	# uploading model
	if(class(model) == "character") {
		# case when model is a path
		body = list('model' = upload_file(model))
	} else {
		# case when model is an object

		# creating temporary file
		saveRDS(model, '.tmp_model')

		# uploading model
		body = list('model' = upload_file('.tmp_model'))

		# setting flag
		del_model = T
	}

	# uploading train dataset
	if(class(train_dataset) == "character" && !grepl("/", train_dataset)) {
		# case when train_dataset is a hash of already uploaded dataset
		body[['train_dataset_hash']] = train_dataset
	} else if(class(train_dataset) == "character") {
		# case when train_dataset is a path to dataset
		#body = c(body, list('train_dataset' = upload_file(train_dataset)))

		body[['train_dataset']] <- upload_file(train_dataset)

		#body['traindataset'] = upload_file
		body[['train_dataset_hash']] = 0
	} else {
		# case when train_dataset is a matrix

		# creating temporary file
		write.table(train_dataset, './tmp_train_data_csv', col.names=T, row.names=F, sep=',')

		# uploading dataset
		body[['train_dataset']] = upload_file('./tmp_train_data_csv')
		body[['train_dataset_hash']]= 0

		# setting flag
		del_train_data = T
	}

	# uploading train_dataset
	if(is.na(test_dataset)) {
		# case when test_dataset is not provided
		body[['is_test_dataset']] = 0
		body[['test_dataset_hash']] = 0
	}
	else if(class(test_dataset) == "character" && !grepl("/", test_dataset)) {
		# case when test_dataset is a hash of already uploaded dataset
		body[['test_dataset_hash']] = test_dataset
		body[['is_test_dataset']] = 1
	}
	else if(class(test_dataset) == "character") {
		# case when test_dataset is a path to csv file
		body[['test_dataset']] = upload_file(test_dataset)
		body[['is_test_dataset']] = 1
		body[['is_test_dataset_hash']] = 0
	}
	else {
		# case when test_dataset is a matrix

		# creating temporary file
		write.table(test_dataset, './tmp_train_data_csv', col.names=F, sep=',')

		# uploading dataset
		body[['test_dataset']] = upload_file('tmp_test_data_csv')

		# setting flag
		del_test_data = T
	}

	# uploading requirements file
	body[['requirements']] = upload_file('.tmp-requirements.txt')

	# uploading sessionInfo
	body[['is_sessionInfo']] = 1
	saveRDS(ses, '.tmp-ses')
	body[['sessionInfo']] = upload_file('.tmp-ses')

	body[['model_name']] = model_name
	body[['system']] = system
	body[['system_release']] = system_release
	body[['distribution']] = distribution
	body[['distribution_version']] = distribution_version
	body[['language']] = language
	body[['language_version']] = language_version
	body[['architecture']] = architecture
	body[['processor']] = processor

	body[['train_data_name']] = train_dataset_name
	body[['test_data_name']] = test_dataset_name

	POST(url = 'http://192.168.137.64/models/post', body = body)

	# removing temporary files
	if(del_model) {
		file.remove('.tmp_model')
	}
	if(del_train_data) {
		file.remove('./tmp_train_data_csv')
	}
	if(del_test_data) {
		file.remove('./tmp_test_data_csv')
	}
}

model_info = function(model_name) {
	content(GET(url = paste0('http://192.168.137.64/models/', model_name, '/info')))
}
